ironfish-zkp/src/circuits/spend.rs

```diff
@@ -50,6 +54,127 @@ pub struct Spend {
     pub sender_address: Option<SubgroupPoint>,
 }
 
+impl Spend {
+    pub fn write<W: Write>(&self, mut writer: W) -> std::io::Result<()> {
+        if let Some(value_commitment) = &self.value_commitment {
+            writer.write_u8(1)?;
+            writer.write_all(&value_commitment.to_bytes())?;
+        } else {
+            writer.write_u8(0)?;
+        }
+        if let Some(proof_generation_key) = &self.proof_generation_key {
+            writer.write_u8(1)?;
+            writer.write_all(proof_generation_key.to_bytes().as_ref())?;
+        } else {
+            writer.write_u8(0)?;
+        }
+        if let Some(payment_address) = &self.payment_address {
+            writer.write_u8(1)?;
+            writer.write_all(payment_address.to_bytes().as_ref())?;
+        } else {
+            writer.write_u8(0)?;
+        }
+        if let Some(commitment_randomness) = &self.commitment_randomness {
+            writer.write_u8(1)?;
+            writer.write_all(commitment_randomness.to_bytes().as_ref())?;
+        } else {
+            writer.write_u8(0)?;
+        }
+        if let Some(ar) = &self.ar {
+            writer.write_u8(1)?;
+            writer.write_all(ar.to_bytes().as_ref())?;
+        } else {
+            writer.write_u8(0)?;
+        }
+        writer.write_all((self.auth_path.len() as u32).to_le_bytes().as_ref())?;
+        let bytes = self
+            .auth_path
+            .iter()
+            .flat_map(|auth_path| {
+                let mut res = vec![];
+                match auth_path {
+                    Some((val, flag)) => {
+                        res.push(1);
+                        res.extend(val.to_bytes_le());
+                        let flag = if *flag { 0u8 } else { 1u8 };
+                        res.push(flag);
+                    }
+                    None => {
+                        res.push(0);
+                    }
+                }
+                res
+            })
+            .collect::<Vec<u8>>();
+        writer.write_all(bytes.as_ref())?;
```

Comment:
This can potentially perform several allocations, which are not necessary. A traditional `for` loop can avoid that:

```rust
for auth_path in &self.auth_path {
    match auth_path {
        Some((val, flag)) => { writer.write_u8(1)?; writer.write_all(val.to_bytes_le())?; writer.write_u8(*flag as u8); }
        None => writer.write_u8(0)?;
    }
}
```